### DISCUSSION TITLE
Removal of metadata for sharing.

### DISCUSSION THREAD

**Discussion by illuminazionetech** [ente-io/ente#4497] [Enhancements]
It would be nice to integrate an option that asks if we want to remove the metadata of the photo before sharing it. It would be an option broken by the community because it would increase privacy. Have a nice day.

**Comment by taivlam** [comment]
I am also interested in this happening in some shape, fashion, or form.
Also, #6105 seems to be a duplicate of this request for enhancement.

**Comment by ghost** [comment]
I totally agree. When sharing a photo or an album, the same we can toggle the allow Download, it would be nice to disable the EXIF and MAP for the viewer. Another cool feature would be hide the original photo/video name and instead show Picture0001 or something generic. Media names often contain Date and Time information.

**Comment by chetbox** [comment]
I think this is really important. Most users do not think about their photos having location data added automatically. This means it

In my case I have been sharing photos with people with a shared album during a long trip. I was not aware I have been sharing my exact location until it was brought to my attention by a friend. My assumption was I am only sharing the content of each photo, not the metadata. The fact this is information is available on a public URL concerns me too.

The metadata data is hidden from users in the ℹ️ menu so it's not very obvious to the person sharing the photos that their location is being shared. To anyone wanting to track this person there is an Open Street Map embedded there.

I think all metadata should be stripped from shared photos/videos by default. Users should be able to opt on to sharing metadata if that's helpful.

**Comment by carleeno** [comment]
+1 on this, this was a feature I used in google photos to hide location of photos when sharing albums publicly (using shared links), I'm very surprised that a privacy-centric service like ente has overlooked this.

**Comment by carleeno** [comment]
So I started doing some digging over the weekend to understand what it would take to add this feature, and unfortunately I realized that e2ee makes this extremely difficult to implement properly, so I figured I'd see if I can stir this discussion up about possible options to implement to maybe get some feedback from the @ente-io  team too.

The main issue is that much of the metadata is embedded in the image/video file itself (Exif and the video container equivalents), and because these files are encrypted blobs until the client (including the web view for public links) decrypts it, it's not possible for the server to strip the metadata from these files.

What this means is that there are a few options that I can see to implement a feature to hide metadata in public links:

1. Implement the "Hide Metadata" feature as a toggle which instructs the web client to not display the metadata in the info pane (or remove the info pane entirely) when the toggle is enabled by the owner.
    - **Pro**: This is fairly easy to implement, and will follow the existing code patterns of the "Allow Downloads" toggle to set an option flag and inform the public web view to modify the UI slightly.
    - **Con**: This doesn't actually prevent the metadata from existing on the viewer's browser, so anybody could still theoretically view it (one could modify the client code locally to ignore the flag, or even just download the file to disk and view the file properties).  
    - **Note**: This has precedence in that the allow downloads toggle doesn't actually prevent the user from downloading the file, it just removes the button to download it, and likewise has a warning when setting the option that it doesn't actually prevent users from downloading the file using external tools.
2. Implement a process to strip the album's files of metadata client-side from the owner's device when they enable the "Hide Metadata" option. Once enabling the option from an album, the client app will begin to process all images and videos, creating a medium-quality version (to save storage space) that contains no embedded metadata. Upon viewing the public link, the server would enforce that only the stripped files are served, and the separate metadata files are not served at all.
    - **Pro**: if implemented correctly, it would not be possible for a viewer of the public link to extract the metadata from the files at all, even if intercepted in the browser and decrypted in a modified client. And the server would refuse to provide minted urls for the original files when the request relates to a collection with this option set.
    - **Con**: this will increase storage cost as every photo and video in a shared album with this feature enabled has to be duplicated to include a stripped version (side-note, the existing streamable video copies do not strip the metadata, so they can't be used unless they are re-processed to strip them too)
    - **Con**: this will incur a large client-side processing delay and bandwidth cost for large albums, and shared links with this feature won't be viewable until processing is complete (or maybe they are partially viewable by only serving the stripped files as they become available). There would have to be a progress UI implemented per album when this toggle is enabled so the owner can tell once the processing has finished
    - **Con**: Implementation for this in code is staggeringly large, as it touches every part of the stack and will be tricky to ensure backwards compatibility, db migration, and bug squashing to ensure no metadata leaks through
    - **Con**: It wouldn't be feasible to pick and choose which metadata is hidden, it'd be all or nothing, as allowing specific metadata filtering would blow up the implementation cost (and if done per-album, storage cost) tremendously
    - **Note**: this does have precedence in the streamable videos option, but would expand to include images, and I'm not sure if the Ente team would be willing to swallow the additional storage costs like they are while streamable videos is in beta.
3. Restructure how files are stored in Ente in general, to fully separate metadata from the image/video data, so metadata can be selectively served.
    - **Pro**: This doesn't bloat storage costs, as there's still only 1 version of the image data stored
    - **Con**: Pretty much a redesign and rewrite of how the entire stack uploads, stores, and serves files; this makes option 2's implementation look like a walk in the park
    - **Con**: Client side processing increases for uploads, as the files have to be stripped of metadata before upload
    - **Con**: Migration and backwards compatibility is downright a nightmare in this version, migration includes every single ente user client having to fully re-process and re-upload all images and videos, not reasonably feasible
    - **Con**: Data export from ente becomes more complicated as it would have to re-join the metadata with the files, and this means that an export would never return the truly original files that were uploaded in the first place, metadata might be altered or missing in case of bugs

These are the options I came up with, but maybe there are more. However of these 3, the only one that seems feasible to be approved by Ente (and the only one I'd be willing to spend time implementing) is the 1st option.

But my concern is that if we implement a superficial "hide metadata" feature that doesn't prevent a slightly savvy user to still view the image location or other metadata, that the feature will be overly trusted and potentially put more user's privacy at risk if the feature isn't understood. Personally, I wouldn't use a superficial option to hide metadata for several of my albums, I just would continue to ensure I don't share anything publicly has that my home/family locations or my current location while traveling, so this version of the hide-metadata feature wouldn't be useful to me, which also discourages me from wanting to implement it that way.

So this is why I figured I'd stop at this point of my research and see if others (especially anybody from Ente ) have any ideas or feedback.
